File: makeips.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#define u8 unsigned char


typedef struct _patch { 
    u8 offset[3];
    u8 size[2];
    u8* bytes;
} patch;

int main(int argc, char *argv[]) {
    if (argc < 3) {
        printf("Give 2 files as arguments");
        return 1; // Not enough arguments
    }

    FILE *f;
    f = fopen(argv[1], "rb");
    if (!f) return 1; // Error opening file
    fseek(f, 0, SEEK_END);
    long base_size = ftell(f);
    fseek(f, 0, SEEK_SET);
    unsigned char *base = (unsigned char *)malloc(base_size);
    fread(base, 1, base_size, f);
    fclose(f);

    f = fopen(argv[2], "rb");
    if (!f) {
        free(base);
        return 1; // Error opening file
    }
    fseek(f, 0, SEEK_END);
    long cmp_size = ftell(f);
    fseek(f, 0, SEEK_SET);
    unsigned char *cmp = (unsigned char *)malloc(cmp_size);
    fread(cmp, 1, cmp_size, f);
    fclose(f);

    int changes_capacity = 100;
    int changes_size = 0;
    patch* changes = (patch *)malloc(changes_capacity * sizeof(patch));

    int i = 0;
    if (base_size <= cmp_size) { // file 2 is bigger
        while (i < base_size) {
            if (base[i] != cmp[i]) {
                int ofs = i;
                int o_size = 0;
                while ((base[i] != cmp[i]) && (i < cmp_size)) {
                    o_size++;
                    i++;
                }
                if (changes_size >= changes_capacity) {
                    changes_capacity += 100;
                    changes = (patch*)realloc(changes, changes_capacity * sizeof(patch));
                }
                changes[changes_size].offset[0] = (ofs & 0xff0000) >> 16;
                changes[changes_size].offset[1] = (ofs & 0xff00) >> 8;
                changes[changes_size].offset[2] = (ofs & 0xff);
                changes[changes_size].size[0] = (o_size & 0xff00) >> 8;
                changes[changes_size].size[1] = (o_size & 0xff);
                changes[changes_size].bytes = (u8*)malloc(o_size * sizeof(u8));
                for(int b = 0; b < o_size; b++){
                    changes[changes_size].bytes[b] = cmp[i - o_size + b]; // hmm
                }
                changes_size++;
            }
            i++;
        }
        
        if(base_size < cmp_size){
            int add_size = 0;
            int ro = i;
            while (i < cmp_size) {
                add_size++;
                i++;
            }
            if (changes_size >= changes_capacity) {
                changes_capacity *= 2;
                changes = (patch*)realloc(changes, changes_capacity * sizeof(patch));
            }
            changes[changes_size].offset[0] = (ro & 0xff0000) >> 16; // Offset
            changes[changes_size].offset[1] = (ro & 0xff00) >> 8;
            changes[changes_size].offset[2] = (ro & 0xff);
            changes[changes_size].size[0] = (add_size & 0xff00) >> 8; // Length
            changes[changes_size].size[1] = (add_size & 0xff);
            changes[changes_size].bytes = (u8*)malloc(add_size * sizeof(u8));
            for(int b = 0; b < add_size; b++){
                changes[changes_size].bytes[b] = cmp[i - add_size + b]; // hmm
            }
            changes_size++;
        }
        
    } 
    else { // file 1 is bigger
        printf("File size mismatch: Ensure file #2 is the larger.");
        return 2;
    }
    
    // length will be 5 + (len of changes * 5) + len of each byte arr 
    changes = (patch *)realloc(changes, changes_size * sizeof(patch));
    int tot = 5 + (5 * changes_size);
    
    //printf("%d, %d\n", changes_size, tot);
    for(int p = 0; p < changes_size; p++) {
        int len = (changes[p].size[0] << 8) | (changes[p].size[1]);
        tot += len;
    }
    
    unsigned char * ips = (u8*)malloc(sizeof(u8)*tot);
    int ips_size = 0;

    ips[ips_size++] = 'P';
    ips[ips_size++] = 'A';
    ips[ips_size++] = 'T';
    ips[ips_size++] = 'C';
    ips[ips_size++] = 'H';

    for (i = 0; i < changes_size; i++) {
        ips[ips_size++] = (u8)(changes[i].offset[0]);
        ips[ips_size++] = (u8)(changes[i].offset[1]);
        ips[ips_size++] = (u8)(changes[i].offset[2]);
        int len = (changes[i].size[0] << 8) | (changes[i].size[1]);
        ips[ips_size++] = (u8)(changes[i].size[0]);
        ips[ips_size++] = (u8)(changes[i].size[1]);
        //ips[ips_size++] = (unsigned char)(changes[i][2]);
        for(int j = 0; j < len; j++) {
            ips[ips_size++] = (u8)changes[i].bytes[j];
        }
    }

    ips[ips_size++] = 'E';
    ips[ips_size++] = 'O';
    ips[ips_size++] = 'F';

    f = fopen("out.ips", "wb");
    fwrite(ips, 1, ips_size, f);
    fclose(f);
    printf("IPS file written\n");

    free(base);
    free(cmp);
    for(int i = 0; i < changes_size; i++){
        free(changes[i].bytes);
    }
    free(changes);

    return 0;
}

File: makeips.py
import os,sys
f = open(sys.argv[1], "rb")
base = f.read()
f.close()
f = open(sys.argv[2], "rb")
cmp = f.read()
f.close()
changes=[]
i = 0
if(len(base) < len(cmp)): # file 1 is bigger 
    while i < len(base):
        if(base[i] != cmp[i]):
            ofs = i
            o = []
            while(base[i] != cmp[i]):
                o.append(cmp[i])
                i += 1
            changes.append((ofs, len(o), o))
        i += 1
    add=[]
    while i < len(cmp):
        add.append(cmp[i])
        i += 1
    changes.append((ofs, len(add), add))
else: # file 2 is bigger 
    while i < len(cmp):
        if(base[i] != cmp[i]):
            ofs = i
            o = []
            while(base[i] != cmp[i]):
                o.append(base[i])
                i += 1
            changes.append((ofs, len(o), o))
        i += 1
    add=[]
    while i < len(base):
        add.append(base[i])
        i += 1
    changes.append((ofs, len(add), add))
i = 0
tot = 0
while i < len(changes):
    tot += changes[i][1]
    i += 1
print(len(changes),"changes found, totaling", tot,"bytes")
ips = []
ips.append(ord('P'))
ips.append(ord('A'))
ips.append(ord('T'))
ips.append(ord('C'))
ips.append(ord('H'))
for c in changes:
    # c[0] offset 3
    # c[1] length 2
    # c[2] bytes  n 
    ips.append((c[0] & 0xff0000) >> 16)
    ips.append((c[0] & 0xff00) >> 8)
    ips.append((c[0] & 0xff))
    ips.append((c[1] & 0xff00) >> 8)
    ips.append((c[1] & 0xff))
    for b in c[2]:
        ips.append(b)
ips.append(ord('E'))
ips.append(ord('O'))
ips.append(ord('F'))
f = open("out.ips", "wb")
f.write(bytes(ips))
f.close()
print("IPS file written")